### ISSUE TITLE
rework on quest tracking

### ISSUE THREAD

**Issue by Javran** [KC3Kai/KC3Kai#1170]
Despite #1117, there are still problems: if user choose to complete their Bw6 (Defeat 12 bosses in the World 4) on 4-4, we will have problem tracking it.

I feel the code related to quest tracking:

https://github.com/KC3Kai/KC3Kai/blob/5787ec93b497dd7bbdd294d442c16f2852609cec/src/library/modules/Kcsapi.js#L1528-L1607

is a bit hard to understand, and using straightforward if-statements will be better for both understanding and maintenance.

I'd like to ask if there are any advantages of using current structure (looping+switching)?
if not I can assign myself working on this to make it look better. (and solve problems like #541 once and for all)

**Comment by dragonjet** [comment]
agreed with conversion to ifs, haven't seem this block before though

**Comment by ReiFan49** [comment]
https://github.com/KC3Kai/KC3Kai/blob/dev-api/src/library/modules/Kcsapi.js#L1639-L1693

well, grouping them using switch-while combo just to make them a bit neater

**Comment by Javran** [comment]
just discussed with @ReiFan49 on gitter channel. another alternative is #918 

@ReiFan49 is more in favor of switch-looping and #918 and I'd recommend #1173 but is also good with #918 (with more  comments to explain). wondering what do you guys think?

**Comment by ReiFan49** [comment]
and also, this need to be solved first before merging the `dev-ledger` branch though

**Comment by dragonjet** [comment]
wait, is ecma6 safe? I know we are already recommendinh Chrome 45+ but are there anything else we need to look out for? I heard something about babeljs that lets you use these safely in a bit older versions of browsers.

**Comment by dragonjet** [comment]
can give sample code on how you guys are planning to apply ecma6 into the quest tracking. I'm starting to feel alright with it since as mentioned, we're already requiring chrome 45+. seems this is a blocker for #1104 which is near `merge-ready` state

**Comment by Javran** [comment]
we can break the dependency, I suggest keeping the quest tracking part intact and merge rest of the changes from `dev-ledger` unless `dev-ledger` branch has been well-tested on quest tracking (reminder: 4-4 should count to Bw6 tracking).

because quest tracking is one feature players would need every day, we have to be more careful.
since it's during maintenance, we have no way to verify.

**Comment by dragonjet** [comment]
this is included in the current v27 release right?

**Comment by ReiFan49** [comment]
according to the dev-ledger previous state
- it's not ECMA6 anymore
- i rushed the commit so forgot the isPvP handling and anything (i don't recall if it still there right now after force commiting right now)
- two alternatives that i suggested is either based **cannot fall-through** on javascript so basically the *while-loop* is used to workaround off the *fall-through* that should be performed instead (but travis said it's illegal to do fall through) and the other one to simplify anything about quest tracking as it does reduces the later work for sortie condition and such

**Comment by dragonjet** [comment]
@Javran @ReiFan49 
I'm proposing a new json file (`quest_reqs.json`) that will look like this:
```
{
    "226": {
        "boss": true,
        "maps": [ [3,3], [3,4], [3,5] ],
        "req_index": 0,
        "min_rank": 3
    },
    ....
}
```
We just make a function that will read this for a "function-call" approach

**Comment by Javran** [comment]
@dragonjet thanks dj, I'm good with this approach.

**Comment by sinsinpub** [comment]
there is another attribute of quest: `period` may need to managed by external json?
currently it's managed by `QuestManager.js`, and 3 arrays for daily, weekly, monthly inside the JS.

**Comment by dragonjet** [comment]
yep, we could add these too.

also when I created the json file, it was reminiscent of the `quests.json` inside each language in TL folders. now I'm thinking we could probably put it there too. or move the `tracking` from TL json into this new one, so other languages than `en` will not mistake to copy them too.

**Comment by sinsinpub** [comment]
yes, those attributes along with quests (`unlock`, `tracking`, `code`, etc) which not related with multi-language could be moved to new data source.
ofc it's nice if the new json could be online updated without updating the webstore extension.

**Comment by dragonjet** [comment]
although code might still be needed there also in the translation jsons, it's their only identifier on what they're trying to translate.

except that if they always use EN json as basis, they just translate the quest `name` and `desc` directly, they don't even need to know the quest code (if that's the case for all TLors then yes we can remove `code`).

we might need some insight from the usual @KC3Kai/translators on what they do.
